src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp

```diff
@@ -44,105 +44,146 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
+        }
+      }
     }
-  }
 
-  for (size_t i = 0; i < dataset.n_cols; ++i)
-  {
-    const double m = std::max(minClusterDistances(assignments[i]),
-                              lowerBounds(i));
-
-    // First bound test.
-    if (upperBounds(i) <= m)
+    #pragma omp critical
     {
-      ++hamerlyPruned;
-      newCentroids.col(assignments[i]) += dataset.col(i);
-      ++counts(assignments[i]);
-      continue;
+      for (size_t i = 0; i < centroids.n_cols; ++i)
+      {
+        minClusterDistances(i) = std::min(minClusterDistances(i), localMinClusterDistances(i));
+      }
     }
+  }
 
-    // Tighten upper bound.
-    upperBounds(i) = distance.Evaluate(dataset.col(i),
-                                       centroids.col(assignments[i]));
-    ++distanceCalculations;
+  const size_t numPoints = dataset.n_cols;
+  const size_t numClusters = centroids.n_cols;
 
-    // Second bound test.
-    if (upperBounds(i) <= m)
-    {
-      newCentroids.col(assignments[i]) += dataset.col(i);
-      ++counts(assignments[i]);
-      continue;
-    }
+  #pragma omp parallel
+  {
+    arma::mat localNewCentroids(centroids.n_rows, centroids.n_cols, arma::fill::zeros);
+    arma::Col<size_t> localCounts(centroids.n_cols, arma::fill::zeros);
+    size_t localHamerlyPruned = 0;
+    size_t localDistanceCalculations = 0;
 
-    // The bounds failed.  So test against all other clusters.
-    // This is Hamerly's Point-All-Ctrs() function from the paper.
-    // We have to reset the lower bound first.
-    lowerBounds(i) = DBL_MAX;
-    for (size_t c = 0; c < centroids.n_cols; ++c)
+    #pragma omp for schedule(static)
+    for (size_t i = 0; i < numPoints; ++i)
     {
-      if (c == assignments[i])
+      const double m = std::max(minClusterDistances(assignments[i]),
+                                lowerBounds(i));
+
+      // First bound test.
+      if (upperBounds(i) <= m + eps)
+      {
+        ++localHamerlyPruned;
+        localNewCentroids.col(assignments[i]) += dataset.col(i);
+        ++localCounts(assignments[i]);
         continue;
+      }
 
-      const double dist = distance.Evaluate(dataset.col(i), centroids.col(c));
+      // Tighten upper bound.
+      upperBounds(i) = distance.Evaluate(dataset.col(i),
+                                         centroids.col(assignments[i]));
+      ++localDistanceCalculations;
 
-      // Is this a better cluster?  At this point, upperBounds[i] = d(i, c(i)).
-      if (dist < upperBounds(i))
+      // Second bound test.
+      if (upperBounds(i) <= m + eps)
       {
-        // lowerBounds holds the second closest cluster.
-        lowerBounds(i) = upperBounds(i);
-        upperBounds(i) = dist;
-        assignments[i] = c;
+        localNewCentroids.col(assignments[i]) += dataset.col(i);
+        ++localCounts(assignments[i]);
+        continue;
       }
-      else if (dist < lowerBounds(i))
+
+      // The bounds failed. So test against all other clusters.
```

Comment:
The original comment said significantly more than this for a reason. Please don't remove comments unless there is a very good reason.